### PR TITLE
Add bytes API.

### PR DESCRIPTION
**Pull request by toots**
This PR adds an API for tracking position in bytes. It should be opt-in and backward compatible.

Fixes: #139

### CODE REVIEW

File: src/lib/sedlexing.ml [hunk 1]
```diff
@@ -21,24 +21,38 @@ type apos = int
...
+      in the input stream *)
+  mutable bytes_pos : int;
+  (* pos is the index, in uchar, in the buffer *)
   mutable curr_bol : int;
-  (* bol is the index in the input stream but not buffer *)
+  (* pos is the index, in bytes, in the buffer *)
+  mutable curr_bytes_bol : int;
+  (* bol is the index, in uchar, in the input stream but not buffer *)
   mutable curr_line : int;
-  (* start from 1, if it is 0, we would not track postion info for you *)
```
Comment (hhugo) [inline src/lib/sedlexing.ml:35]: I would keep this comment

Comment (toots) [reply src/lib/sedlexing.ml:35]: It should have been removed with https://github.com/ocaml-community/sedlex/commit/7581a9d8a05ca82f8d37b9f6a107ffe8e08b854b, gonna remove it now.

File: src/lib/sedlexing.ml [hunk 2]
```diff
@@ -21,24 +21,38 @@ type apos = int
...
       in the input stream *)
   mutable pos : int;
-  (* pos is the index in the buffer *)
+  (* Position of the first byte in buffer
+      in the input stream *)
+  mutable bytes_pos : int;
+  (* pos is the index, in uchar, in the buffer *)
   mutable curr_bol : int;
-  (* bol is the index in the input stream but not buffer *)
+  (* pos is the index, in bytes, in the buffer *)
```
[the remark this answers is not in the document]
Comment (toots) [reply src/lib/sedlexing.ml:39]: I think that it is correct, the diff is confusing tho. It does apply to `curr_bytes_bol`, the counterpart to `curr_bol` with the same comment.

**Comment by toots** [comment]
PR is confirmed working with liquidsoap and also confirmed that we were reporting bogus positions before, as we assumed the returned positions were in bytes not code points.

File: src/lib/sedlexing.ml [hunk 3]
```diff
@@ -6,6 +6,14 @@ exception InvalidCodepoint of int
 exception MalFormed
 
 module Uchar = struct
+  (* This for compatibility with ocaml < 4.14.0 *)
+  let utf_8_byte_length _ = 1
```
Comment (hhugo) [inline src/lib/sedlexing.ml:10]: I think you should just duplicate the small function from the stdlib

File: src/lib/sedlexing.ml (hunk 1 above)
Comment (hhugo) [reply src/lib/sedlexing.ml]: I didn't realize you've completely removed the feature. You should update the doc comment of `set_position` then.

```
(** set the initial tracked input position for [lexbuf].
        If set to [Lexing.dummy_pos], Sedlexing will not track position
        information for you. *)
val set_position : lexbuf -> Lexing.position -> unit
```

File: src/lib/sedlexing.ml (hunk 3 above)
Comment (hhugo) [reply src/lib/sedlexing.ml:10]: For the record, here are the two functions 
```ocaml
let utf_8_byte_length u = match to_int u with
| u when u < 0 -> assert false
| u when u <= 0x007F -> 1
| u when u <= 0x07FF -> 2
| u when u <= 0xFFFF -> 3
| u when u <= 0x10FFFF -> 4
| _ -> assert false

let utf_16_byte_length u = match to_int u with
| u when u < 0 -> assert false
| u when u <= 0xFFFF -> 2
| u when u <= 0x10FFFF -> 4
| _ -> assert false


```

File: test/basic.ml [hunk 4]
```diff
@@ -142,43 +230,97 @@ let%expect_test "utf8" =
   [%expect
```
Comment (hhugo) [inline test/basic.ml:230]: It'd be nice to update the utf8 test so that the input contains uchar spanning multiple bytes.

File: test/basic.ml [hunk 5]
```diff
@@ -1,20 +1,36 @@
 let digit = [%sedlex.regexp? '0' .. '9']
 let number = [%sedlex.regexp? Plus digit]
 
+let print_pos buf =
+  let f { Lexing.pos_lnum; pos_bol; pos_cnum; _ } =
+    Printf.sprintf "line=%d:bol=%d:cnum=%d" pos_lnum pos_bol pos_cnum
```
Comment (hhugo) [inline test/basic.ml:6]: a `line:column` format could be easier to read.
and matybe `line:column-line:column` like in error messages.

File: src/lib/sedlexing.ml (hunk 3 above)
Comment (toots) [reply src/lib/sedlexing.ml]: Done thks!

File: test/basic.ml (hunk 5 above)
Comment (toots) [reply test/basic.ml:6]: I'm trying to report the actual internal parameters for easier debugging.

File: test/basic.ml (hunk 4 above)
Comment (toots) [reply test/basic.ml:230]: Done!

**Comment by toots** [comment]
@hhugo I think that this one is ready for merge!

**Comment by hhugo** [comment]
> @hhugo I think that this one is ready for merge!

I still think comments are weird. See my comments https://github.com/ocaml-community/sedlex/pull/146/files#r1244060086

File: src/lib/sedlexing.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (toots) [reply src/lib/sedlexing.ml:60]: Haa ok I see, yes no pb.

**Comment by hhugo** [comment]
Also see my reply in https://github.com/ocaml-community/sedlex/pull/146#discussion_r1244047393

File: src/lib/sedlexing.ml (hunk 1 above)
Comment (toots) [reply src/lib/sedlexing.ml]: I did this:
```ocaml
(** set the initial tracked input position, in code point, for [lexbuf].
        If unspecified, byte postion is set to the same value as code
        point position. *)
 ```

I cannot see anything in the code disabling position tracking when passing `dummy_pos`, which is only defined abstractly in the [official doc](https://v2.ocaml.org/api/Lexing.html), meaning any expectation on internal values could be broken by future releases, so I just removed it.

**Comment by toots** [comment]
> Also see my reply in [#146 (comment)](https://github.com/ocaml-community/sedlex/pull/146#discussion_r1244047393)

Addressed!

**Comment by toots** [comment]
> > @hhugo I think that this one is ready for merge!
> 
> I still think comments are weird. See my comments https://github.com/ocaml-community/sedlex/pull/146/files#r1244060086

Addressed!

File: src/lib/sedlexing.ml [hunk 6]
```diff
@@ -70,17 +114,24 @@ let empty_lexbuf =
...
+  }
 
-let set_position lexbuf position =
+let set_position ?bytes_position lexbuf position =
   lexbuf.offset <- position.Lexing.pos_cnum - lexbuf.pos;
   lexbuf.curr_bol <- position.Lexing.pos_bol;
-  lexbuf.curr_line <- position.Lexing.pos_lnum
+  lexbuf.curr_line <- position.Lexing.pos_lnum;
+  let bytes_position = Option.value ~default:position bytes_position in
+  lexbuf.bytes_offset <- bytes_position.Lexing.pos_cnum - lexbuf.pos;
```
Comment (hhugo) [inline src/lib/sedlexing.ml:129]: should this be  `lexbuf.bytes_offset <- bytes_position.Lexing.pos_cnum - lexbuf.bytes_pos;` ?

Comment (toots) [inline src/lib/sedlexing.ml:129]:
```suggestion
  lexbuf.bytes_offset <- bytes_position.Lexing.pos_cnum - lexbuf.bytes_pos;
```

**Comment by hhugo** [comment]
LGTM

**Comment by hhugo** [comment]
Also fixes https://github.com/ocaml-community/sedlex/issues/96

**Comment by hackwaly** [comment]
So, can we use utf16 positions as well? Language server protocol currently only support utf16 positions

**Pull request merged**